### DISCUSSION TITLE
Starting Keycloak with CRDB (Cockroach DB)

### DISCUSSION THREAD

**Discussion by briankahoot** [keycloak/keycloak#15277] [Archive]
I have been trying to run keycloak 19 and now 20 using a cockroach db deployment but it is always failing on startup as part of the liquibase migration. I have found some threads about this but can't quite figure out how to get it working. Currently the error I'm getting is:

```
 Caused by: org.postgresql.util.PSQLException: ERROR: relation "user_required_action" (362): unimplemented: primary key dropped without subsequent addition of new primary key in same transaction
```

Can someone help me get this working? Do I need to run the database migrations separately from the server startup or are there specific server settings that need to be set to run with cockroach db?

**Comment by hmlnarik** [comment]
cc: @sguilhen @ahus1 @vramik

**Reply by ahus1** [reply]
@briankahoot - looking at the error message it seems to me that you're running Keycloak with the legacy store. See https://www.keycloak.org/2022/07/storage-map to read more legacy vs. map store.

The legacy doesn't support CockroachDB. The new map store will eventually support CockroachDB. The support in KC20 for CockroachDB in a nutshell: Keycloak will start, will initialize the DB, and you'll be able to log in. Still there are some integration test failures we're currently working on; see #11618. With a lot of them fixed in KC20, those remaining issues might or might not affect you. 

To try out the new map store with CockroachDB, read the blog post listed above and follow the instructions in the FAQ at #9596.

**Reply by briankahoot** [reply]
That was it... I had tried specifying the storage as jpa before but turns out I had done it incorrectly. The server has now started!

I'll have a closer look at the failures listed on that ticket but I don't think there is anything that will stop us from at least testing how we'd like to use the server.

Thank you for the quick response.
